src/extensions.rs

```diff
@@ -59,12 +59,12 @@ impl Extensions {
     /// assert!(ext.insert(4u8).is_none());
     /// assert_eq!(ext.insert(9i32), Some(5i32));
     /// ```
-    pub fn insert<T: Send + Sync + 'static>(&mut self, val: T) -> Option<T> {
+    pub fn insert<T: Clone + Send + Sync + 'static>(&mut self, val: T) -> Option<T> {
```

Comment:
This breaks tonic interceptors that attach extensions with well-defined ownership relying on `&mut self`.

My tonic interceptors that attach internal metrics and internal metadata now have to have something like `Arc<Mutex<Option<MyExtension>>>`.
Before:
```rust
let mut extension = request.extensions_mut().remove().expect("extension was set");
```

After:
```rust
let extension: Arc<Mutex<Option<ReusableOwned<Detailer>>>> = request
    .extensions_mut()
    .remove()
    .unwrap_or_default();
let extension = std::mem::take(&mut *extension.lock().expect("one locker"));
let mut extension = extension.expect("extension was set");
```

Reply:
Yes, this is expected. It's a trade-off, and we settled on merging this.